enhancements/network/cloud-egress-ip.md

```diff
@@ -0,0 +1,603 @@
+   an assignment and issues a cloud egress IP assignment request by creating a
+   `cloudprivateipconfig.cloud.openshift.io` CR with `.spec.items` filled out.
+4. Watches for the update of the CR and inspects `.status.items` to
+   retrieve the assignment status.
+5. Performs cluster level networking setup for the egress IPs.  
+6. Triggers an `event` to the user notifying successful/unsuccessful assignment.
+
+### Risks and Mitigations
+
+#### Risks
+
+Finalizing the work for 4.8 has a risk arising mainly CI/CD setup for this new
+component.
+
+Given future OpenShift features requiring cloud level modifications of the
+network fabric (once such existing future enhancement is: node port range
+modifications): the proposal here is to call the repository;
+openshift/cloud-network-config-controller. The name would explicitly indicate
+that this component performs cloud level network modifications, without digging
+itself into a "egress IP dedicated hole". New features would be added and
+supported by defining separate CRDs.
+
+##### Node CSR renewal
+
+Clusters deployed on clouds such as the ones described in this document,
+retrieve the node addresses from the cloud provider API. The following links
+show the process employed for each:
+
+[AWS]
+
+AWS seems to use all IPs listed in the instance's metadata:
+`network/interfaces/macs/$mac/local-ipv4` as the node's `InternalIP`. The
+metadata will also be populated with egress IP once the assignment has been
+performed, see: [the AWS section](#aws). The node is thus expected to have the
+egress IP listed as the node's `InternalIP` once the IP has been attached. This
+might have impacts on the CSR renewal for the node.
+
+[Azure]
+
+Azure, seems to only grab the first private IP addresses listed for the primary
+NIC associated with the instance. This means that the egress IP assignment,
+should not show up in the node's `InternalIP` addresses.
+
+[GCE]
+
+GCE iterates over all NICs associated with the instance and retrieves the IP
+address associated with each to set the node's `InternalIP` addresses. This
+means that the egress IP assignment will not show up among the node's
+`InternalIP` addresses, because the egress IP will be assigned as an "alias IP",
+as described in [the GCP section](#gcp)
+
+[AWS]: https://github.com/kubernetes/legacy-cloud-providers/blob/ce3a9094e0ca26c7c4676be32b292e9836affade/aws/aws.go#L1498-L1502
+[Azure]: https://github.com/kubernetes/legacy-cloud-providers/blob/ce3a9094e0ca26c7c4676be32b292e9836affade/azure/azure_instances.go#L122-L127
+[GCE]: https://github.com/kubernetes/legacy-cloud-providers/blob/ce3a9094e0ca26c7c4676be32b292e9836affade/gce/gce_instances.go#L107-L117
+
+The Kubelet will issue a CSR renewal for the IPs known to it. On node reboot on
+AWS it would come to pick up the new egress IP, subsequently causing issues if
+the machine approver does not sync its expected IPs which are pending CSR
+approval with the cloud API.
```

Comment:
So this all ties into the long-running "WTF is Node.Status.Addresses actually _for_?" argument, but I feel like we don't want egress IPs to show up in Node.Status.Addresses. As @michaelgugino pointed out in https://bugzilla.redhat.com/show_bug.cgi?id=1860774#c15, they are not addresses that can be used to reach the node. And also, they're explicitly transient, which makes it even less likely that you can come up with a valid reason why putting them into Node.Status.Addresses would be helpful to users.

Unfortunately, as noted in the discussion above, there doesn't seem to be any way to add an IP to an AWS or GCP node _without_ having it show up in Node.Status.Addresses. And in fact, there doesn't seem to be any way to add any extra metadata to an IP address object in AWS/GCP, so there's not even any possible way that the AWS/GCP CloudProviders _could_ distinguish "real" from "transient"/"internal use only" node IPs.

So it seems like if we want to support people adding extra IPs to their cloud nodes (either for the egress IP feature, or for any other future feature, or for customer-internal use cases), then we need to relax CSR approval to allow unexpected extra IPs...